### ISSUE TITLE
Memory allocation/free bugs

### ISSUE THREAD

**Issue by adefazio** [tqchen/xgboost#172]
There appears to be some memory related issues causing problems for me. In particular I'm getting the error:
 
     the error is *** glibc detected *** xgboost: free(): invalid next size (normal): 0x0000000000777e80 ***

It doesn't occur on all machines I run it on, even with the same data, so it's likely that memory layout is causing an intermittent effect.

It occurs for task=pred runs in particular. Running valgrind over the program might be the best way for you to debug this. Unfortunately I can't provide the data I'm using. The configuration file is:

``` booster=gbtree
min_child_weight=1
gamma=1.0
save_period=10
data="train_control.libsvm"
eval[train]="train_control.libsvm"
model_dir="models-2015-02-10T23-40-28"
model_out="models-2015-02-10T23-40-28/latest.model"
######################################
max_depth=1
eta=0.1
num_round=256
objective=reg:linear
eval[test]=test_control.libsvm
```

The command is:

xgboost "configuration.conf" \
    task=pred model_in="latest.model" \
    test:data="test_random.libsvm" \
    name_pred="test_random-model-control.pred"

The output from valgrind is:
```
==7003== Memcheck, a memory error detector
==7003== Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
==7003== Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
==7003== Command: xgboost configuration.conf task=pred model_in=latest.model test:data=test_random.libsvm name_pred=test_random-model-control.pred
==7003== 
70361x783 matrix with 9780179 entries is loaded from test_random.libsvm.buffer
start prediction...
==7003== Invalid write of size 4
==7003==    at 0x406AB7: xgboost::tree::RegTree::FVec::Fill(xgboost::SparseBatch::Inst const&) (model.h:506)
==7003==    by 0x4396AF: xgboost::gbm::GBTree::Pred(xgboost::SparseBatch::Inst const&, long, int, unsigned int, xgboost::tree::RegTree::FVec*, float*, unsigned long, unsigned int) (gbtree-inl.hpp:325)
==7003==    by 0x438826: xgboost::gbm::GBTree::Predict(xgboost::IFMatrix*, long, xgboost::BoosterInfo const&, std::vector<float, std::allocator<float> >*, unsigned int) (gbtree-inl.hpp:158)
==7003==    by 0x45C459: xgboost::learner::BoostLearner::PredictRaw(xgboost::learner::DMatrix const&, std::vector<float, std::allocator<float> >*, unsigned int) const (learner-inl.hpp:404)
==7003==    by 0x45C11E: xgboost::learner::BoostLearner::Predict(xgboost::learner::DMatrix const&, bool, std::vector<float, std::allocator<float> >*, unsigned int, bool) const (learner-inl.hpp:338)
==7003==    by 0x45E712: xgboost::BoostLearnTask::TaskPred() (xgboost_main.cpp:253)
==7003==    by 0x45CCA4: xgboost::BoostLearnTask::Run(int, char**) (xgboost_main.cpp:71)
==7003==    by 0x4532B3: main (xgboost_main.cpp:319)
==7003==  Address 0x5ef3d24 is 0 bytes after a block of size 2,868 alloc'd
==7003==    at 0x4C285FC: operator new(unsigned long) (vg_replace_malloc.c:298)
==7003==    by 0x415114: __gnu_cxx::new_allocator<xgboost::tree::RegTree::FVec::Entry>::allocate(unsigned long, void const*) (new_allocator.h:89)
==7003==    by 0x412E08: std::_Vector_base<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::_M_allocate(unsigned long) (stl_vector.h:140)
==7003==    by 0x410012: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::_M_fill_insert(__gnu_cxx::__normal_iterator<xgboost::tree::RegTree::FVec::Entry*, std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> > >, unsigned long, xgboost::tree::RegTree::FVec::Entry const&) (vector.tcc:414)
==7003==    by 0x40C9E5: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::insert(__gnu_cxx::__normal_iterator<xgboost::tree::RegTree::FVec::Entry*, std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> > >, unsigned long, xgboost::tree::RegTree::FVec::Entry const&) (stl_vector.h:851)
==7003==    by 0x40A622: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::resize(unsigned long, xgboost::tree::RegTree::FVec::Entry) (stl_vector.h:557)
==7003==    by 0x406A2E: xgboost::tree::RegTree::FVec::Init(unsigned long) (model.h:500)
==7003==    by 0x4385FC: xgboost::gbm::GBTree::Predict(xgboost::IFMatrix*, long, xgboost::BoosterInfo const&, std::vector<float, std::allocator<float> >*, unsigned int) (gbtree-inl.hpp:134)
==7003==    by 0x45C459: xgboost::learner::BoostLearner::PredictRaw(xgboost::learner::DMatrix const&, std::vector<float, std::allocator<float> >*, unsigned int) const (learner-inl.hpp:404)
==7003==    by 0x45C11E: xgboost::learner::BoostLearner::Predict(xgboost::learner::DMatrix const&, bool, std::vector<float, std::allocator<float> >*, unsigned int, bool) const (learner-inl.hpp:338)
==7003==    by 0x45E712: xgboost::BoostLearnTask::TaskPred() (xgboost_main.cpp:253)
==7003==    by 0x45CCA4: xgboost::BoostLearnTask::Run(int, char**) (xgboost_main.cpp:71)
==7003== 
==7003== Invalid write of size 4
==7003==    at 0x406B14: xgboost::tree::RegTree::FVec::Drop(xgboost::SparseBatch::Inst const&) (model.h:512)
==7003==    by 0x439838: xgboost::gbm::GBTree::Pred(xgboost::SparseBatch::Inst const&, long, int, unsigned int, xgboost::tree::RegTree::FVec*, float*, unsigned long, unsigned int) (gbtree-inl.hpp:336)
==7003==    by 0x438826: xgboost::gbm::GBTree::Predict(xgboost::IFMatrix*, long, xgboost::BoosterInfo const&, std::vector<float, std::allocator<float> >*, unsigned int) (gbtree-inl.hpp:158)
==7003==    by 0x45C459: xgboost::learner::BoostLearner::PredictRaw(xgboost::learner::DMatrix const&, std::vector<float, std::allocator<float> >*, unsigned int) const (learner-inl.hpp:404)
==7003==    by 0x45C11E: xgboost::learner::BoostLearner::Predict(xgboost::learner::DMatrix const&, bool, std::vector<float, std::allocator<float> >*, unsigned int, bool) const (learner-inl.hpp:338)
==7003==    by 0x45E712: xgboost::BoostLearnTask::TaskPred() (xgboost_main.cpp:253)
==7003==    by 0x45CCA4: xgboost::BoostLearnTask::Run(int, char**) (xgboost_main.cpp:71)
==7003==    by 0x4532B3: main (xgboost_main.cpp:319)
==7003==  Address 0x5ef3d24 is 0 bytes after a block of size 2,868 alloc'd
==7003==    at 0x4C285FC: operator new(unsigned long) (vg_replace_malloc.c:298)
==7003==    by 0x415114: __gnu_cxx::new_allocator<xgboost::tree::RegTree::FVec::Entry>::allocate(unsigned long, void const*) (new_allocator.h:89)
==7003==    by 0x412E08: std::_Vector_base<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::_M_allocate(unsigned long) (stl_vector.h:140)
==7003==    by 0x410012: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::_M_fill_insert(__gnu_cxx::__normal_iterator<xgboost::tree::RegTree::FVec::Entry*, std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> > >, unsigned long, xgboost::tree::RegTree::FVec::Entry const&) (vector.tcc:414)
==7003==    by 0x40C9E5: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::insert(__gnu_cxx::__normal_iterator<xgboost::tree::RegTree::FVec::Entry*, std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> > >, unsigned long, xgboost::tree::RegTree::FVec::Entry const&) (stl_vector.h:851)
==7003==    by 0x40A622: std::vector<xgboost::tree::RegTree::FVec::Entry, std::allocator<xgboost::tree::RegTree::FVec::Entry> >::resize(unsigned long, xgboost::tree::RegTree::FVec::Entry) (stl_vector.h:557)
==7003==    by 0x406A2E: xgboost::tree::RegTree::FVec::Init(unsigned long) (model.h:500)
==7003==    by 0x4385FC: xgboost::gbm::GBTree::Predict(xgboost::IFMatrix*, long, xgboost::BoosterInfo const&, std::vector<float, std::allocator<float> >*, unsigned int) (gbtree-inl.hpp:134)
==7003==    by 0x45C459: xgboost::learner::BoostLearner::PredictRaw(xgboost::learner::DMatrix const&, std::vector<float, std::allocator<float> >*, unsigned int) const (learner-inl.hpp:404)
==7003==    by 0x45C11E: xgboost::learner::BoostLearner::Predict(xgboost::learner::DMatrix const&, bool, std::vector<float, std::allocator<float> >*, unsigned int, bool) const (learner-inl.hpp:338)
==7003==    by 0x45E712: xgboost::BoostLearnTask::TaskPred() (xgboost_main.cpp:253)
==7003==    by 0x45CCA4: xgboost::BoostLearnTask::Run(int, char**) (xgboost_main.cpp:71)
==7003== 
writing prediction to test_random-model-control.pred
==7003== 
==7003== HEAP SUMMARY:
==7003==     in use at exit: 0 bytes in 0 blocks
==7003==   total heap usage: 1,242 allocs, 1,242 frees, 82,056,699 bytes allocated
==7003== 
==7003== All heap blocks were freed -- no leaks are possible
==7003== 
==7003== For counts of detected and suppressed errors, rerun with: -v
==7003== ERROR SUMMARY: 44 errors from 2 contexts (suppressed: 6 from 6)

**Comment by tqchen** [comment]
Thanks so much for reporting the issue!
I am guessing that this might be caused by  additional feature index that occurs on test set(which greater than those in training set).

I have pushed a fix on this, can you try and see if it fixed the problem?

**Comment by adefazio** [comment]
Thanks for the speedy reply, the issue appears to be fixed with that update.
